goamiddleware/jwt_middleware.go

```diff
@@ -0,0 +1,284 @@
+			if err != nil {
+				goa.LogError(ctx, err.Error())
+				return ErrJWTError(err)
+			}
+
+			requiredScopes := goa.ContextRequiredScopes(ctx)
+
+			for _, scope := range requiredScopes {
+				if !scopesInClaim[scope] {
+					msg := "authorization failed: required 'scope' or 'scopes' not present in JWT claim"
+					return ErrJWTError(msg, "required", requiredScopes, "scopes", scopesInClaimList)
+				}
+			}
+
+			ctx = goajwt.WithJWT(ctx, token)
+			if validationFunc != nil {
+				nextHandler = validationFunc(nextHandler)
+			}
+			return nextHandler(ctx, rw, req)
+		}
+	}
+}
+
+func extractTokenFromHeader(schemeName string, req *http.Request) (string, error) {
+	val := req.Header.Get(schemeName)
+	if val == "" {
+		return "", ErrJWTError(fmt.Sprintf("missing header %q", schemeName))
+	}
+
+	if !strings.HasPrefix(strings.ToLower(val), "bearer ") {
+		return "", ErrJWTError(fmt.Sprintf("invalid or malformed %q header, expected 'Bearer JWT-token...'", val))
+	}
+
+	incomingToken := strings.Split(val, " ")[1]
+
+	return incomingToken, nil
+}
+
+func extractTokenFromQueryParam(schemeName string, req *http.Request) (string, error) {
+	incomingToken := req.URL.Query().Get(schemeName)
+	if incomingToken == "" {
+		return "", ErrJWTError(fmt.Sprintf("missing parameter %q", schemeName))
+	}
+
+	return incomingToken, nil
+}
+
+// validScopeClaimKeys are the claims under which scopes may be found in a token
+var validScopeClaimKeys = []string{"scope", "scopes"}
+
+// parseClaimScopes parses the "scope" or "scopes" parameter in the Claims. It
+// supports two formats:
+//
+// * a list of strings
+//
+// * a single string with space-separated scopes (akin to OAuth2's "scope").
+//
+// An empty string is an explicit claim of no scopes.
+func parseClaimScopes(token *jwt.Token) (map[string]bool, []string, error) {
+	scopesInClaim := make(map[string]bool)
+	var scopesInClaimList []string
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		return nil, nil, fmt.Errorf("unsupport claims shape")
+	}
+	for _, k := range validScopeClaimKeys {
+		if rawscopes, ok := claims[k]; ok && rawscopes != nil {
+			switch scopes := rawscopes.(type) {
+			case string:
+				for _, scope := range strings.Split(scopes, " ") {
+					scopesInClaim[scope] = true
+					scopesInClaimList = append(scopesInClaimList, scope)
+				}
+			case []interface{}:
+				for _, scope := range scopes {
+					if val, ok := scope.(string); ok {
+						scopesInClaim[val] = true
+						scopesInClaimList = append(scopesInClaimList, val)
+					}
+				}
+			default:
+				return nil, nil, fmt.Errorf("unsupported scope format in incoming JWT claim, was type %T", scopes)
+			}
+			break
+		}
+	}
+	sort.Strings(scopesInClaimList)
+	return scopesInClaim, scopesInClaimList, nil
+}
+
+// ErrJWTError is the error returned by this middleware when any sort of validation or assertion
+// fails during processing.
+var ErrJWTError = goa.NewErrorClass("jwt_security_error", 401)
```

Comment:
We could use goajwt.ErrJWTError instead of redeclaring it.

Reply:
"could" -> "should"